userguide: update what is Kill Bill doc (#327)
killbill/killbill-docs: userguide/tutorials/what_is_kill_bill.adoc

== Introduction

> Kill Bill is an open-source programmable platform that enables you to build custom billing solutions. That's a mouthful, to say the least, but think of it as building blocks to implement your own billing workflows.

Reviewer: ```suggestion
Kill Bill is an open-source, programmable platform that enables you to build custom billing solutions. Think of it as building blocks to implement your own billing workflows.
```

> Similar to the Unix philosophy of composability over monolithic design, Kill Bill provides APIs to create various pricing models, manage subscriptions, generate invoices, integrate with various payment gateways, etc. Whether you are selling access to a digital service for $10 a month or selling a multi-tier cell-phone data plan, Kill Bill provides the primitives to do it so you don't have to start from scratch.

Reviewer: ```suggestion
Kill Bill APIs let you create various pricing models, manage subscriptions, generate invoices, integrate with various payment gateways, etc. Whether you are selling access to a digital service for $10 a month or selling a multi-tier cell phone data plan, Kill Bill provides the primitives to do it so you don't have to start from scratch.
```

Author: The Unix philosophy bit would speak to the engineering audience. Do you feel strongly against it?

> _Does Kill Bill support per CPU-hour billing?_ is the wrong kind of question to ask when evaluating Kill Bill. This would be similar to asking _Does Spring Boot offer endpoints to store user data?_: Spring Boot gives you all the tools you need to create endpoints which can take in user data and let you persist that data in a database, without having to worry about parsing HTTP headers or setting up database connection pooling, but it doesn't offer a user data endpoint. This is similar in the case of Kill Bill: while it doesn't understand _CPU hour billing_, you can create a usage-based plan in the catalog, which is billed based on the number of consumed _CPU-hour_ units.

Reviewer: I think this is going to lose a lot of people who don't know what Spring Boot is (hand raised). It's really just a repetition of the previous paragraphs, so I recommend deleting it. ([Kill your darlings](https://www.masterclass.com/articles/what-does-it-mean-to-kill-your-darlings)!)

Author: Fair enough. That being said, _bis repetita placent_ 😁

Can you think of another metaphor we could use here, to drive home that point? This is an aspect often misunderstood about Kill Bill.

> While Kill Bill is used for various cases (b2b, b2c, offline, etc.), to get a better sense of what Kill Bill offers out of the box, the easiest is to look at how it is used in a typical b2c e-commerce environment.

Reviewer: ```suggestion
While Kill Bill is used for various cases (B2B, B2C, offline, etc.), to get a better sense of what Kill Bill offers out of the box, let's look at how it is used in a typical B2C e-commerce environment.
```

> The *subscription billing* features allow an online store to create recurring or usage based subscriptions (e.g. "gold plan subscription at $9.99/month") and manage upgrades and downgrades. The system will take care of generating invoices, managing credits and triggering payments. Netflix, Dropbox and The New York Times are examples of websites which sell such subscriptions.

Author: Oxford comma fan, eh? ;-)

> The *payment* features let the store connect with virtually any payment gateway and processor (Stripe, Braintree, PayPal, Adyen, etc.) to charge your customers (using credit or debit cards, ACH, wire transfers, Bitcoin, etc.) for their subscriptions or one-off payments (e.g. shopping cart-type websites like Amazon or Zappos).

Reviewer: ```suggestion
Kill Bill's *payment* features let the online store connect with virtually any payment gateway and processor (Stripe, Braintree, PayPal, Adyen, etc.) to charge customers (using credit or debit cards, ACH, wire transfers, Bitcoin, etc.) for their subscriptions or one-off payments (e.g., shopping cart-type websites like Amazon or Zappos).
```

Check this video for a quick tour:

> https://www.youtube.com/watch?v=f2IHcz3OLYo[image:http://img.youtube.com/vi/f2IHcz3OLYo/0.jpg[align=center]]

Author: Fixed!

== Overview

=== Big Picture

> At a high level, we provide 2 pieces of software that **you will have to deploy and manage** in your own data center, in your private cloud, or in any public cloud.

Reviewer: ```suggestion
At a high level, we provide two pieces of software that **you will have to deploy and manage** in your own data center, in your private cloud, or in any public cloud.
```

Author: Not sure about rules either, but the arrowheads show the flow of HTTP traffic (requests). I think it makes sense to keep them.

You do have a point regarding the other high level graphic missing a key, but I'm not sure what I would put as a legend (it's logically external / internal traffic, but I think calling it that way would be confusing...).

> * Kill Bill, or Kill Bill server, is a standalone Java server that runs in your back-end. Your customers will never interact with it directly, but instead your website (it could be a custom website, a Drupal or WordPress deployment, etc.) and potentially your internal tools (CRM, CPQ, etc.) will trigger REST API calls (over HTTP) to create subscriptions, retrieve invoices, give credits, etc.

Reviewer: ```suggestion
* Kill Bill, or Kill Bill server, is a standalone Java server that runs in your back-end environment. Your customers never interact with Kill Bill directly. Instead, by using your website (it could be a custom website, a Drupal or WordPress deployment, etc.) and potentially your internal tools (CRM, CPQ, etc.), they trigger REST API calls (over HTTP) to create subscriptions, retrieve invoices, make payments, etc.
```

Author: Reads funny to me? Are you sure about the _By using your website and your internal tools, they..._ ?

To be clear, there are two bits here: on one hand the website and on the other the internal tools. Customers will only interact through the website, company employees would be the ones using the internal tools.

> * Kaui, our Back-office UI, which provides an interface to interact with Kill Bill for your customer support, product, and finance teams to manage refunds, invoice adjustments, reporting, etc.

Reviewer: ```suggestion
* Kaui, our back-office user interface, lets your teams (customer support, product, finance, etc.) interact with Kill Bill to manage refunds, invoice adjustments, reporting, and so forth.
```

> * Customers interact with your website, which then issues REST API calls to Kill Bill. **You are responsible for designing and building that piece**.

Reviewer: ```suggestion
* As mentioned earlier, customers interact with your website, which then issues REST API calls to Kill Bill. **You are responsible for designing and building that piece**.
```

* Kill Bill, with its database, **manages all the state associated with the subscriptions, invoices, payments, etc.**

> * Kaui is the **Back-office UI for your teams**, and it includes role based access control to provide different priviledges for different teams/users:

Reviewer: ```suggestion
* Kaui, the **back-office UI for your teams**, includes role-based access control to provide different privileges for the various teams/users:
```

>   ** Your customer support team will use to access/modify consumer accounts data

Reviewer: ```suggestion
  ** Your customer support team uses Kaui to access/modify consumer accounts data.
```

>   ** Your product team  will use to configure catalog information, billing rules, ...

Reviewer: ```suggestion
  ** Your product team uses Kaui to configure catalog information, billing rules, etc.
```

>   ** You finance/accounting team will access it for different types of reports

Reviewer: ```suggestion
  ** Your finance/accounting team access Kaui for different types of reports.
```

>   ** Your engineering team will access it for running quick scenarios for development and testing purposes

Reviewer: ```suggestion
  ** Your engineering team access Kaui for running quick scenarios for development and testing purposes.
```

=== Deployment

> Kill Bill is packaged as a self-contained WAR and is therefore executed within a web container (typically Tomcat). While you can deploy Kill Bill directly as a WAR inside a container, we also provide additional deployment options, such as Docker containers or _1-click button_ AWS AMIs.

Reviewer: ```suggestion
Kill Bill is packaged as a self-contained WAR and is therefore executed within a web container (typically Tomcat). While you can deploy Kill Bill directly as a WAR inside a container, we also provide additional deployment options, such as Docker containers or AWS AMIs with _1-click deployments_.
```

Here is a typical Kill Bill deployment architecture:

[caption=""]
//.Kill Bill: Zoomed-in Picture

> image::../assets/what_is_killbill/KillBill_Mid_level.png[format=png,align=center,title-align=center]

Reviewer 2: This is the **physical** view of the deployment, while the first diagram, the 'high level' is the **logical** view

Reviewer: I would change the olive-green color of the JDBC line to something brighter. Right now it's color is too close to black.

Author: I've renamed the caption to mention *physical* view.

I've also changed the colour to something brighter.

* There are two sets of _servers_, one for Kill Bill, one for Kaui. Each _server_ can either be a physical server, a VM, or a container. Either way, it will run a Tomcat Java process where Kill Bill or Kaui are deployed. Typically, two instances of Kaui and three of Kill Bill should be plenty.
* Two (physical or virtual) load-balancers route traffic to the servers (in a _least number of connections_ or _round robin_ configuration). Assuming the network can be trusted, these load-balancers terminate SSL traffic.
* The Kill Bill and Kaui instances are stateless, but need a relational database. The database server can be shared (but the applications should always use two different schemas) or not.

> In a nutshell, a Kill Bill deployment follows a typical microservices deployment. Common DevOps techniques, such as rolling updates, Blue-Green deployments and Canary deployments, are applicable as well.

Reviewer: ```suggestion
In a nutshell, a Kill Bill deployment follows a typical microservices deployment. Common DevOps techniques, such as rolling updates, Blue-Green deployments, and Canary deployments, are applicable as well.
```
